File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-1.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        maxlength = 0
        seen = set()
        l = 0
        for r in range(len(s)):
            while s[r] in seen:
                seen.remove(s[l]) #start removing from the left
                l+=1 # move pointer towards right
            seen.add(s[r])
            maxlength = max(maxlength, r-l+1)
        return maxlength

        

File: Data Structures & Algorithms/max-water-container/submission-1.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        //area = height * width
        //here area = (right-left+1)*min(height[l], height[r])
        //move the pointer which is shorter
        int maxArea = 0;
        int left = 0, right = heights.size()-1;
        while (left<right){
            int area = (right-left)*min(heights[left], heights[right]);
            maxArea = max(maxArea, area);
            if(heights[left]<=heights[right]){
                left++;
            }
            else{
                right--;
            }
        }
        return maxArea;
    }
};
